README.md
# PA55 in Dart

A reference implementation of PA55 (not PA55 NYAPS) in Dart. Apache [license](https://github.com/pa55/pa55-dart/blob/master/LICENSE).

## Usage -- command line

```
pub run example/pa55_example.dart -h
```
displays the arguments that the program takes as follows. For example,

```
pub run example/pa55_example.dart -a SHA256 -r 10K -l 16 -v
```
runs PA55 with the hash algorithm set to SHA256, the PBKDF2 rounds to 10,000 and generates a password of length 16 after obtaining the inputs from the user through ```stdin``` for the master secret and the hint. Turning off the ```-v``` flag will turn off all output messages from PA55 except the final password and any error.

## Usage of the library

A simple usage example:

```dart
import 'package:pa55/pa55.dart';

main() {
  String pa55Password = PA55v1Core.generatePBKDF2Password(
                        HashAlgorithm.SHA256,
                        "test1234",
                        "1234test",
                        18, //24-character password
                        10000);
}
```

## Features and bugs

Please file feature requests and bugs at the [issue tracker][tracker].

[tracker]: https://github.com/pa55/pa55-dart/issues


CHANGELOG.md
## 1.0.0

- Reference implementation in Dart. See Java reference implementation at https://github.com/pa55/pa55


test/pa55_test.dart
/*
 * Copyright 2019 Anirban Basu

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
 */

import 'package:pa55/pa55.dart';
import 'package:test/test.dart';

void main() {

  group('HMac hash algorithms with test1234, 1234test, 16 characters, 10K iterations', () {

    test('(SHA1)', () {
      expect(PA55v1Core.generatePBKDF2Password(HashAlgorithm.SHA1,
          "test1234", "1234test", 12, 10000), "M5sfFklaZ86upPyp");
    });

    test('(SHA256)', () {
      expect(PA55v1Core.generatePBKDF2Password(HashAlgorithm.SHA256,
          "test1234", "1234test", 12, 10000), "arJngCkH1meNNDhN");
    });

    test('(SHA512)', () {
      expect(PA55v1Core.generatePBKDF2Password(HashAlgorithm.SHA512,
          "test1234", "1234test", 12, 10000), "FIbqvrcQy1EJAZVG");
    });

  });

  group('PBKDF2 iterations with SHA1, abcd1234, 1234abcd, 24 characters', () {

    test('(10K iterations)', () {
      expect(PA55v1Core.generatePBKDF2Password(HashAlgorithm.SHA1,
          "abcd1234", "1234abcd", 18, 10000), "oM/vrSqxpqVWJtlovMZsEDVV");
    });

    test('(250K iterations)', () {
      expect(PA55v1Core.generatePBKDF2Password(HashAlgorithm.SHA1,
          "abcd1234", "1234abcd", 18, 250000), "c+qFBNPQe0hxVOzYbp2lBi0G");
    });

    test('(500K iterations)', () {
      expect(PA55v1Core.generatePBKDF2Password(HashAlgorithm.SHA1,
          "abcd1234", "1234abcd", 18, 500000), "M/cv4059AVG4K2deMGm+J/3C");
    }, timeout: Timeout.factor(2));

  });

}


lib/src/pa55v1_core.dart
/*
 * Copyright 2019 Anirban Basu

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
 */

import 'dart:convert';
//TODO: optimise the imports
import 'package:pointycastle/export.dart';

enum HashAlgorithm {
  SHA1,
  SHA256,
  SHA512
}

class PA55v1Core {

  static String generatePBKDF2Password(HashAlgorithm pbkdfAlgorithm,
                                String masterSecret,
                                String passwordHint,
                                int passwordLength,
                                int pbkdf2Rounds) {
    var digest;
    var blockLength;
    switch(pbkdfAlgorithm) {
      /*
       * Check block lengths from:
       * https://github.com/bcgit/bc-java/blob/master/core/src/main/java/org/bouncycastle/crypto/macs/HMac.java
       */
      case HashAlgorithm.SHA1:
        digest = SHA1Digest();
        blockLength = 64;
        break;
      case HashAlgorithm.SHA256:
        digest = SHA256Digest();
        blockLength = 64;
        break;
      case HashAlgorithm.SHA512:
        digest = SHA512Digest();
        blockLength = 128;
        break;
    }

    //UTF-8 encoder: string->bytes
    Utf8Encoder utf8encoder = Utf8Encoder();

    //Generator for PBE derived keys and ivs as defined by PKCS 5 V2.0 Scheme 2
    PBKDF2KeyDerivator derivator = PBKDF2KeyDerivator(
        HMac(digest, blockLength))
          ..init(Pbkdf2Parameters(utf8encoder.convert(passwordHint),
          pbkdf2Rounds, passwordLength));

    //Obtain the PBKDF2 output
    var bytes = derivator.process(utf8encoder.convert(masterSecret));

    //Return a base64 encoded string
    Base64Encoder base64encoder = Base64Encoder();
    return base64encoder.convert(bytes);
  }

}


example/pa55_example.dart
/*
 * Copyright 2019 Anirban Basu

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
 */

import 'dart:io';
import 'package:args/args.dart';
import 'package:pa55/pa55.dart';

void main(List<String> arguments) {
  final stopwatch = Stopwatch();
  var exitCode = 0; //presume success
  final programVersion = 'PA55 v.2019.10';
  final parser = ArgParser();

  parser..addOption('hash', abbr: 'a', help: 'Hash algorithm.',
                  allowed: {'SHA1', 'SHA256', 'SHA512'},
                  defaultsTo: 'SHA256');
  parser.addOption('length', abbr: 'l', help: 'Desired password length in characters.',
                  allowed: {'8', '12', '16', '20', '24', '28', '32'},
                  defaultsTo: '12');
  parser.addOption('rounds', abbr: 'r', help: 'PKBDF2 rounds (high values will be slower)',
                  allowed: {'10K', '250K', '500K', '750K', '1M', '1.25M', '1.5M'},
                  defaultsTo: '500K');
  parser.addFlag('verbose', abbr: 'v', help: 'En(dis)able verbose output.',
                negatable: true, defaultsTo: false);
  parser.addFlag('help', abbr: 'h', help: 'Show this usage information.',
                negatable: false);
 try {
    var argResults = parser.parse(arguments);
    if(argResults.wasParsed('help') || argResults.arguments.isEmpty) {
      printUsage(parser, programVersion);
    }
    else {
      var verbose = argResults['verbose'];
      //master secret and the password hint must be parsed from stdin instead of
      // being passed as an argument
      stdin.echoMode = false; //disable echo for sensitive input
      if(verbose) {
        print(programVersion);
        stdout.write("Type the master secret and press ENTER: ");
      }
      var masterSecret = stdin.readLineSync(retainNewlines: false);
      if(verbose) {
        stdout.write("\r\nType the password hint and press ENTER: ");
      }
      var passwordHint = stdin.readLineSync(retainNewlines: false);
      stdin.echoMode = true; //enable echo
      if(verbose) {
      //print the options
        print('\r\nComputing password, using options');
        argResults.options.forEach((f) => {
          if(f!='help' && f!='verbose') { //no need to show these flags
            print('  ' + f.toString() + ': ' + argResults[f].toString())
          }
        });
        stopwatch..start();
      }
      String pa55Password = PA55v1Core.generatePBKDF2Password(
                            hashAlgorithmFromString(argResults['hash']),
                            masterSecret,
                            /*argResults['hint']*/passwordHint,
                            passwordLengthFromString(argResults['length']),
                            passwordRoundsFromString(argResults['rounds']));
      if(verbose) {
        var timeElapsed = stopwatch.elapsedMilliseconds;
        print("Password (computed in " + timeElapsed.toString() + " ms): " + pa55Password);
        stopwatch.stop();
      }
      else {
        print(pa55Password);
      }
    }
  } on Exception catch (e) {
    stderr.write(e.toString() + '\r\n');
    printUsage(parser, programVersion);
    exitCode = -1;
  }
  exit(exitCode);
}

void printUsage(ArgParser parser, String programVersion) {
  print(programVersion + ' usage information');
  print('IMPORTANT: The master secret and the password hint must be typed in or piped in from another source. These cannot be passed as arguments.');
  print(parser.usage);
}

HashAlgorithm hashAlgorithmFromString(String argument) {
  HashAlgorithm choice;
  switch(argument) {
    case "SHA1":
      choice = HashAlgorithm.SHA1;
      break;
    case "SHA256":
      choice = HashAlgorithm.SHA256;
      break;
    case "SHA512":
      choice = HashAlgorithm.SHA512;
      break;
    default:
      choice = null;
  }
  return choice;
}

num passwordLengthFromString(String length) {
  return ((int.tryParse(length)?? 0)*3)~/4;
}

num passwordRoundsFromString(String rounds) {
  var thousandUnits = rounds[rounds.length - 1];
  var numberPart = rounds.substring(0, rounds.length-1);
  num parsedNumber = (double.tryParse(numberPart)?? 0);
  switch(thousandUnits.toUpperCase()) {
    case 'K':
      parsedNumber *= 1000;
    break;
    case 'M':
      parsedNumber *= 1000000;
    break;
  }
  return parsedNumber.toInt();
}
